### ISSUE TITLE
Performance (accuracy) of a Transformer model is somewhat below expectations

### ISSUE THREAD

**Issue by gsoul** [OpenNMT/OpenNMT-tf#40]
I trained several models for En <-> Fr on WMT14 data, and I get around 30 BLEU with these models. While Google's paper says they achieved ~42 BLEU. I'm aware that such high BLEU is dependent on batch-size/LR/number of GPUs for training, but my expectations were still to get somewhere around 38 BLEU.

So here is my set-up:
1. All the EN <-> Fr data from WMT 14 concatenated together and shuffled. It's about 40 mln sentence pairs.
2. Learnt BPE of size 32k
3. Src and Tgt dictionaries are both of size 32k tokens.
4. All the parameters are from default config
5. ~315k steps with batch-size of 128 (maximum size that fits into 1080ti RAM)

P.S. When I evaluate same model on newstest2013 - I get ~28BLEU

Do you have any insights on what might be imporved here, or if there are some significant differences with T2T implementation?

**Comment by guillaumekln** [comment]
Which direction are you reporting, EN->FR or FR->EN?

**Comment by gsoul** [comment]
Both. I have about the same results for them. Less than 0.5 BLEU difference.

**Comment by JunchengYao** [comment]
I trained a model for En -> Fr on wmt14 data, 1400k steps with batch-size of 2048, 32k tokens.
I evaluate the model on newstest2014 - I get ~38BLEU.
The wmt14 data have 40.8M sentence pairs, but the paper used 36M. Do you have any preprocessing or cleaning procedure before trainning? 
Do you have any preprocessing procedure before BLEU test?

**Comment by guillaumekln** [comment]
Did you renew your experiments?

We made some great progress lately and are now very close in performance to the reference implementation at [tensor2tensor](https://github.com/tensorflow/tensor2tensor); at least on the English-German dataset. See this new [run configuration](https://github.com/OpenNMT/OpenNMT-tf/blob/master/config/transformer_1gpu.yml).

Regarding the English-French dataset, I think it might require additional preprocessing steps to get the 36M sentences dataset referenced in the paper. See for example: https://github.com/facebookresearch/fairseq/issues/59

**Comment by gsoul** [comment]
As for my experiments, they are on pause for a bit, as I'm busy with some immediate tasks at work, though I plan to renew them soon.

In terms of performance and OpenNMT-tf - I watch the repo closely on a daily basis and see your progress on new configs for training as well as adding support for training on multiple GPUs, which looks super-cool. I plan to test it soon. Though EnFr is a bigger mystery for me (I got comparable to the paper results for EnDe with Google's code, but not French: https://github.com/tensorflow/tensor2tensor/issues/518). I'll try filters suggested by FAIR, though not sure if that'd be good enough (never know, until you try). The question about dataset size to T2T team is left unanswered so far: https://github.com/tensorflow/tensor2tensor/issues/507

I'll report here when I get new results.

P.S. I think I may start working on this pretty soon: https://github.com/OpenNMT/OpenNMT-tf/issues/10 . Could you please advise what are the best places to start digging into your code there?

**Comment by gsoul** [comment]
Lukasz Kaizer just replied that they did not do any data preprocessing unlike folks from FAIR: https://github.com/tensorflow/tensor2tensor/issues/507#issuecomment-364627908

The only limit they put on the data - maximum length of a sentence.

**Comment by guillaumekln** [comment]
There is now a script to reproduce good results with the Transformer on the English-German WMT dataset: https://github.com/OpenNMT/OpenNMT-tf/tree/master/scripts/wmt

**Comment by apeterswu** [comment]
Actually, En-De translation can be reproduced. The question is about En-Fr... It still cannot be reproduced.

**Comment by guillaumekln** [comment]
Did you also try this task? What results did you get?
